dhall-lsp-server/src/Dhall/LSP/Handlers.hs

```diff
@@ -2,59 +2,348 @@ module Dhall.LSP.Handlers where
 
 import qualified Language.Haskell.LSP.Core as LSP
 import qualified Language.Haskell.LSP.Messages as LSP
-import qualified Language.Haskell.LSP.Utility as LSP
-
 import qualified Language.Haskell.LSP.Types as J
 import qualified Language.Haskell.LSP.Types.Lens as J
+import qualified Language.Haskell.LSP.VFS as LSP
+import qualified Data.Aeson as J
+import qualified Yi.Rope as Rope
+
+import Dhall.Core (Expr, pretty)
+import Dhall.Parser (Src(..))
+import Dhall.TypeCheck (X)
 
-import qualified Dhall.LSP.Handlers.Diagnostics as Diagnostics
-import qualified Dhall.LSP.Handlers.DocumentFormatting as Formatting
+import Dhall.LSP.Backend.Dhall (FileIdentifier, parse, load, typecheck,
+  normalize, fileIdentifierFromFilePath, fileIdentifierFromURI, invalidate,
+  cacheExpr, parseWithHeader)
+import Dhall.LSP.Backend.Diagnostics (Range(..), Diagnosis(..), explain,
+  sanitiseRange, rangeFromDhall, diagnose)
+import Dhall.LSP.Backend.Formatting (formatExprWithHeader)
+import Dhall.LSP.Backend.Linting (Suggestion(..), suggest, lint)
+import Dhall.LSP.Backend.Typing (typeAt, srcAt, annotateLet)
+import Dhall.LSP.State
 
+import GHC.Conc (atomically)
+import Control.Applicative ((<|>))
+import Control.Concurrent.STM.TVar
 import Control.Lens ((^.))
-import Control.Monad.Reader (runReaderT)
+import Control.Monad.Trans (lift, liftIO)
+import Control.Monad.Trans.Except (throwE, catchE, runExceptT)
+import Control.Monad.Trans.State.Strict (execStateT)
+import Control.Monad.Trans.State.Strict (gets, modify)
+import qualified Data.HashMap.Strict as HashMap
+import qualified Data.Map.Strict as Map
+import Data.Maybe (maybeToList)
+import Data.Text (Text)
+import qualified Data.Text as Text
+import Lens.Family (view, set, over)
+import qualified Network.URI as URI
+import qualified Network.URI.Encode as URI
+import Text.Megaparsec (SourcePos(..), unPos)
 
--- handler that doesn't do anything. Useful for example to make haskell-lsp shut
--- up about unhandled DidChangeTextDocument notifications (which are already
--- handled haskell-lsp itself).
-nullHandler :: LSP.LspFuncs () -> a -> IO ()
-nullHandler _ _ = return ()
 
-{- Currently implemented by the dummy nullHandler:
-initializedHandler :: LSP.LspFuncs () -> J.InitializedNotification -> IO ()
+-- Workaround to make our single-threaded LSP fit dhall-lsp's API, which
+-- expects a multi-threaded implementation.
+wrapHandler
+  :: TVar (Maybe ServerState)
+  -> (a -> HandlerM ())
+  -> a
+  -> IO ()
+wrapHandler vstate handle message = do
+  mstate <- readTVarIO vstate
```

Comment:
I'm not too concerned about race conditions since I don't intend to introduce multitasking at all. The use of `TVar` over `IORef` dates back to the sample code from `haskell-lsp`...

On the other hand, I do like the idea of using `withMVar` to be on the safe side. Will change that!